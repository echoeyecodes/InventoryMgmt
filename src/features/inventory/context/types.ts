import { Inventory } from "../models/Inventory";

export type AddInventoryParams = Omit<
  Inventory,
  "id" | "createdAt" | "updatedAt"
>;
export type UpdateInventoryParams = {
  id: string;
  data: Omit<Partial<Inventory>, "createdAt" | "updatedAt">;
};

export type InventoryContextType = {
  data: Inventory[];
  addInventory: (inventory: AddInventoryParams) => void;
  updateInventory: (inventory: UpdateInventoryParams) => void;
  removeInventory: (id: string) => void;
  getInventory: (id: string) => Inventory | undefined;
  getInventoryByTitle: (title: string) => Inventory | undefined;
};
