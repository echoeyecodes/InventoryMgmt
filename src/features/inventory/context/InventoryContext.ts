import { createContext } from "react";
import { InventoryContextType } from "./types";

export const InventoryContext = createContext<InventoryContextType>({
  data: [],
  addInventory: () => {},
  updateInventory: () => {},
  removeInventory: () => {},
  getInventory: () => undefined,
  getInventoryByTitle: () => undefined,
});
