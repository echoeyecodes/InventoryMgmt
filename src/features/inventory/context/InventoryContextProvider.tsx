import {
  PropsWithChildren,
  useEffect,
  useMemo,
  useState,
} from "react";
import { InventoryContext } from "./InventoryContext";
import { Inventory } from "../models/Inventory";
import storage from "@/lib/storage";
import { AddInventoryParams, UpdateInventoryParams } from "./types";
import { ActivityIndicator } from "react-native";

const INVENTORY_KEY = "inventory_db";

export const InventoryContextProvider = (props: PropsWithChildren) => {
  const [isLoading, setIsLoading] = useState(true);
  const [data, setData] = useState<Inventory[]>([]);

  const addInventory = (inventory: AddInventoryParams) => {
    const now = new Date().toISOString();
    const newData: Inventory[] = [
      ...data,
      {
        id: now,
        createdAt: now,
        updatedAt: now,
        ...inventory,
      },
    ];
    setData(newData);
    storage.setItem(INVENTORY_KEY, JSON.stringify(newData));
  };

  const removeInventory = (id: string) => {
    const newData = data.filter((inventory) => inventory.id !== id);
    setData(newData);
    storage.setItem(INVENTORY_KEY, JSON.stringify(newData));
  };

  const updateInventory = (inventory: UpdateInventoryParams) => {
    const now = new Date().toISOString();
    const newData = data.map((item) =>
      item.id === inventory.id
        ? { ...item, ...inventory.data, updatedAt: now }
        : item
    );
    setData(newData);
    storage.setItem(INVENTORY_KEY, JSON.stringify(newData));
  };

  const getInventory = (id: string) => {
    return data.find((inventory) => inventory.id === id);
  };

  const getInventoryByTitle = (title: string) => {
    return data.find(
      (inventory) =>
        inventory.title.trim().toLowerCase() === title.trim().toLowerCase()
    );
  };

  useEffect(() => {
    const fetchData = async () => {
      setIsLoading(true);
      const data = await storage.getItem(INVENTORY_KEY);
      setData(data ? JSON.parse(data) : []);
      setIsLoading(false);
    };
    fetchData();
  }, []);

  const sortedData = useMemo(() => {
    const sortedData = data.sort((a: Inventory, b: Inventory) => {
      return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime();
    });
    return sortedData;
  }, [data]);

  return (
    <InventoryContext.Provider
      value={{
        data: sortedData,
        getInventoryByTitle,
        addInventory,
        updateInventory,
        removeInventory,
        getInventory,
      }}
    >
      {isLoading ? <ActivityIndicator /> : props.children}
    </InventoryContext.Provider>
  );
};
