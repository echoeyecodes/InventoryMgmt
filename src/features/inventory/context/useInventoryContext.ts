import { useContext } from "react";
import { InventoryContext } from "./InventoryContext";

export const useInventoryContext = () => {
  return useContext(InventoryContext);
};
