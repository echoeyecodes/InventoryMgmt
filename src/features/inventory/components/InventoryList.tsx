import { IconButton } from "@/components/button";
import { DefaultLayout } from "@/components/display/DefaultLayout";
import { ThemedText } from "@/components/typography/ThemedText";
import {
  ListRenderItem,
  StyleSheet,
  TouchableOpacity,
  View,
} from "react-native";
import AntDesign from "@expo/vector-icons/AntDesign";
import { Divider } from "@/components/Divider";
import { InventoryItem } from "@/features/inventory/components/InventoryItem";
import { FlatList } from "@/components/list/FlatList";
import { useCallback } from "react";
import { Inventory } from "@/features/inventory/models/Inventory";
import { spacing } from "@/constants/spacing";
import { useInventoryContext } from "@/features/inventory/context/useInventoryContext";
import { useTranslation } from "@/lib/i18n";
import { EmptyInventory } from "@/features/inventory/components/EmptyInventory";
import { useThemeColor } from "@/lib/hooks/colors/useThemeColor";
import { useRouter } from "expo-router";

export default function InventoryList() {
  const router = useRouter();
  const colors = useThemeColor();
  const { data } = useInventoryContext();
  const { t } = useTranslation();

  const onInventoryPress = (id: string) => {
    router.push(`/add-inventory?id=${id}`);
  };

  const renderItem: ListRenderItem<Inventory> = useCallback(
    ({ item }) => (
      <TouchableOpacity
        onPress={() => onInventoryPress(item.id)}
        testID={`inventory-item`}
      >
        <InventoryItem data={item} />
      </TouchableOpacity>
    ),
    []
  );

  const navigateToAddInventory = () => {
    router.push("/add-inventory");
  };

  return (
    <DefaultLayout style={styles.container} testID="inventory-list">
      <View style={styles.header}>
        <ThemedText variant="h1">{t("common.home")}</ThemedText>
        <IconButton
          onPress={navigateToAddInventory}
          testID="add-inventory-button"
        >
          <AntDesign name="pluscircle" size={24} color={colors.text} />
        </IconButton>
      </View>

      <Divider />

      <FlatList
        data={data}
        renderItem={renderItem}
        style={styles.flatList}
        contentContainerStyle={styles.flatListContent}
        ListEmptyComponent={<EmptyInventory />}
      />
    </DefaultLayout>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  header: {
    padding: spacing.md,
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
  },
  flatList: {
    flex: 1,
    paddingVertical: spacing.md,
  },
  flatListContent: {
    gap: spacing.md,
    paddingHorizontal: spacing.md,
    paddingBottom: spacing.xl,
  },
});
