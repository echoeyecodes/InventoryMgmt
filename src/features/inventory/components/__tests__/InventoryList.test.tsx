import React from "react";
import { render, waitFor, fireEvent } from "@testing-library/react-native";
import { InventoryContextProvider } from "../../context/InventoryContextProvider";
import InventoryList from "../InventoryList";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { NavigationContainer } from "@react-navigation/native";

jest.mock("@react-native-async-storage/async-storage", () => ({
  setItem: jest.fn(),
  getItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn(),
}));

const mockPush = jest.fn();
jest.mock("expo-router", () => ({
  useRouter: () => ({
    push: mockPush,
  }),
}));

describe("<InventoryList />", () => {
  it("matches snapshot", async () => {
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(JSON.stringify([]));

    const { toJSON } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <InventoryList />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it("renders correctly", async () => {
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(JSON.stringify([]));

    const { getByTestId, getByText } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <InventoryList />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      expect(getByTestId("inventory-list")).toBeTruthy();
      expect(getByText("Home")).toBeTruthy();
      expect(getByTestId("add-inventory-button")).toBeTruthy();
      expect(getByTestId("empty-inventory-text")).toBeTruthy();
    });
  });

  it("navigates to AddInventory screen when add button is clicked", async () => {
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(JSON.stringify([]));

    const { getByTestId } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <InventoryList />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      expect(getByTestId("add-inventory-button")).toBeTruthy();
    });

    const addButton = getByTestId("add-inventory-button");
    fireEvent.press(addButton);

    expect(mockPush).toHaveBeenCalledWith("/add-inventory");
  });

  it("shows all inventory items", async () => {
    const mockInventoryItems = [
      { id: "1", title: "Item 1", quantity: 5 },
      { id: "2", title: "Item 2", quantity: 10 },
    ];
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(
      JSON.stringify(mockInventoryItems)
    );

    const { getAllByTestId } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <InventoryList />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      const inventoryItems = getAllByTestId("inventory-item");
      expect(inventoryItems.length).toBe(2);
    });
  });

  it("navigates to AddInventory screen with id when an inventory item is tapped", async () => {
    const mockInventoryItems = [
      { id: "1", title: "Item 1", quantity: 5 },
      { id: "2", title: "Item 2", quantity: 10 },
    ];
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(
      JSON.stringify(mockInventoryItems)
    );

    const { getAllByTestId } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <InventoryList />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      const inventoryItem = getAllByTestId("inventory-item");
      fireEvent.press(inventoryItem[0]);
    });

    expect(mockPush).toHaveBeenCalledWith("/add-inventory?id=1");
  });
});
