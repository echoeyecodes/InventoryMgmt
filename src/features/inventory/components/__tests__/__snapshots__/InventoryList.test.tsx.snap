// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InventoryList /> matches snapshot 1`] = `<ActivityIndicator />`;
